File: lab5/Dictionary.c
/*------------------------------------------------
File Name: Dictionary.c
Desc: A dictionary in C using linked lists
Instructions: Run using DictionaryClient harType <in> <out>

Name:   Joseph Csoti
CruzID: 1617438
Class:  CMPS 12M
Date:   02/20/18
------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

#include "Dictionary.h"

// private types --------------------------------

// NodeObj
typedef struct NodeObj{
  char* key;
  char* value;
  struct NodeObj* next;
} NodeObj;
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* k, char* v) {
  Node N = malloc(sizeof(NodeObj));
  assert(N!=NULL);
  N->key = k;
  N->value = v;
  N->next = NULL;
  return(N);
}

// freeNode()
// destructor for the Node type
void freeNode(Node pN){
    free(pN);
    pN = NULL;
}
// ------------------------------------

//DictionaryObj
typedef struct DictionaryObj {
  struct NodeObj* head;
} DictionaryObj;

// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void) {
  Dictionary D = malloc(sizeof(DictionaryObj));
  assert(D!=NULL);
  D->head = NULL;
  return D;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD) {
  if( pD!=NULL && *pD!=NULL ){
    free(*pD);
    *pD = NULL;
 }
}

// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D) {
  return D->head == NULL;
}

// size()
// returns the number of (key, value) pairs in D
// pre: none
int size(Dictionary D) {
  int size = 0;
  Node cur = D->head;

  while(cur != NULL) {
    size++;
    cur = cur->next;
  }

  return size;
}

// lookup()
// returns the value v such that (k, v) is in D, or returns NULL if no 
// such value v exists.
// pre: none
char* lookup(Dictionary D, char* k) {

  //Get head
  Node cur = D->head;

  while(cur != NULL) {
    if(strcmp(cur->key, k) == 0)
       return cur->value;
    
    cur = cur->next;
  }

  return NULL;
}

// insert()
// inserts new (key,value) pair into D
// pre: lookup(D, k)==NULL
void insert(Dictionary D, char* k, char* v) {

  //Make sure key does not exist
  if(lookup(D, k) != NULL) {
    fprintf(stderr, "Error: key collision\n");
    return;
  }

  //if head node
  if(D->head == NULL) {
    D->head = newNode(k, v);
    return;
  }

  //If in body
  Node cur = D->head;
  while(cur->next != NULL) {
    cur = cur->next;
  }

  cur->next = newNode(k, v);
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL
void delete(Dictionary D, char* k) {

  Node cur = D->head;

  //if head
  if((strcmp(cur->key, k) == 0)) {
    Node copy = D->head;
    D->head = cur->next;
    freeNode(copy);
    return;
  }

  //Find node
  while(cur->next != NULL) {
     if(strcmp(cur->next->key, k) == 0){
       Node copy = cur->next;
       cur->next = cur->next->next;
       freeNode(copy);
       return;
     }
    
    cur = cur->next;
  }
}

// makeEmpty()
// re-sets D to the empty state.
// pre: none
void makeEmpty(Dictionary D) {

  Node head = D->head;
  Node temp;

  //Free each node
  while (head != NULL) {
    temp = head;
    head = head->next;
    freeNode(temp);
  }

  freeDictionary(D);
}

// printDictionary()
// pre: none
// prints a text representation of D to the file pointed to by out
void printDictionary(FILE* out, Dictionary D) {

  Node cur = D->head;

  while(cur != NULL) {
    fprintf(out, "%s %s\n", cur->key, cur->value);
    cur = cur->next;
  }
}

File: lab7/Dictionary.java
/*------------------------------------------------
File Name: Dictionary.java
Desc: Dictionary ADT using a BST structure
Instructions:  Create an instance using Dictionary() in another file

Name:   Joseph Csoti
CruzID: 1617438 // jcsoti
Class:  CMPS 12M
Date:   03/08/18
------------------------------------------------*/

class Dictionary implements DictionaryInterface {

  private int size;
  private Node root;

  // Node class
  private class Node {
    String key, value;
    Node left, right;
    Node(String key, String value) {
      this.key = key;
      this.value = value;
    }
  }

  //constructor
  Dictionary() {
    root = null;
    size = 0;
  }

  // isEmpty()
  // pre: none
  // returns true if this Dictionary is empty, false otherwise
  public boolean isEmpty() {
    return size == 0;
  }

  // size()
  // pre: none
  // returns the number of entries in this Dictionary
  public int size() {
    return size;
  }

  // lookup()
  // pre: none
  // returns value associated key, or null reference if no such key exists
  public String lookup(String key) {
    return lookup(root, key);
  }

  private String lookup(Node parent, String key) {

    if(parent == null)
      return null;

    if(key.equals(parent.key))
      return parent.value;

    if(key.compareTo(parent.key) < 0) 
      return lookup(parent.left, key);
    else
      return lookup(parent.right, key);
  }

  private Node findNode(String key) {
    return findNode(root, key);
  }

  private Node findNode(Node parent, String key) {

    if(parent == null)
      return null;

    if(key.equals(parent.key))
      return parent;

    if(key.compareTo(parent.key) < 0) 
      return findNode(parent.left, key);
    else
      return findNode(parent.right, key);
  }

  private Node findParent(Node root, Node node) {
    Node parent = null;
    if(!root.equals(node)) {
      parent = root;
      while(!node.equals(parent.left) && !node.equals(parent.right)) {
        if(node.value.compareTo(parent.value) < 0)
          parent = parent.left;
        else
          parent = parent.right;
       }
    }
    return parent;
  }

  // findMin()
  // find minimum item in subtree rooted at R
  Node findMin(Node root){
    Node n = root;
    if(n != null){
      for(; n.left != null; n=n.left);
    }
    return n;
  }

  // insert()
  // inserts new (key,value) pair into this Dictionary
  // pre: lookup(key)==null
  public void insert(String key, String value) throws DuplicateKeyException {
    if(lookup(key) != null)
      throw new DuplicateKeyException("insert(): Duplicate key");

    // Create proper nodes
    Node n = new Node(key, value); 
    Node parent = null;
    Node child = root;

    while(child != null) {
      //Walk the tree together...
      parent = child;
      
      // ...until 'C' takes the spot where the new node 'n' would be
      if(key.compareTo(child.key) < 0)
        child = child.left;
      else
        child = child.right;
    }

    // Empty dict, node becomes root
    if(parent == null)
      root = n;
    else if(key.compareTo(parent.key) < 0)
      parent.left = n; //Left child
    else
      parent.right = n; //Right child

    size++;
  }

  // delete()
  // deletes pair with the given key
  // pre: lookup(key)!=null
  public void delete(String key) throws KeyNotFoundException {
    if(lookup(key) == null)
      throw new KeyNotFoundException("delete(): Key not found");
    
    // Init nodes
    Node parent, s, n;

    // Get node OBJ assosiated w/ key
    n = findNode(key);

    // CASE 1: No children
    if(n.left == null && n.right == null) {
      if(n.equals(root)) {
        root = null;
      }
      else {
        parent = findParent(root, n);
        
        if(n.equals(parent.right))
          parent.right = null;
        else
          parent.left = null;
      }
    }
    // CASE 2-A: Has left child
    else if(n.right == null) {
      if(n.equals(root)) {
        root = n.left;
      }
      else {
        parent = findParent(root, n);

        if(n.equals(parent.right))
          parent.right = n.left;
        else
          parent.left = n.left;
      }

      System.out.println("case 2 - no right");
      
    }
    // CASE 2-B: Has right child
    else if(n.left == null) {
      if(n.equals(root)) {
        root = n.right;
      }
      else {
        parent = findParent(root, n);

        if(n.equals(parent.right))
          parent.right = n.right;
        else
          parent.left = n.right;
      }  
    }
    // CASE 3: Has BOTH right and left channel
    else {
      s = findMin(n.right); //get min node from right side of tree
      n.value = s.value;
      parent = findParent(n, s); // get its parent

      if(s.equals(parent.right))
        parent.right = s.right;
      else
        parent.left = s.right; 
    }

    size--;
  }

  // makeEmpty()
  // pre: none
  public void makeEmpty() {
    root = null;
    size = 0;
  }

  // toString()
  // returns a String representation of this Dictionary
  // overrides Object's toString() method
  // pre: none
  public String toString() {
    return print(root);
  }

  //Recursive print function
  private String print(Node parent){
    if(parent == null)
      return "";

    String str, a, b;

    a = print(parent.left); //get left side of tree
    str = parent.key + " " + parent.value + "\n"; // print "head" node
    b = print(parent.right); //get right side of tree

    return a + str + b; //combine
 }
}